core/src/serve_repair.rs

```diff
@@ -170,6 +167,95 @@ struct ServeRepairStats {
     err_id_mismatch: usize,
 }
 
+struct RepairBudget {
+    last_update: Instant,
+    epoch_staked_nodes: Option<Arc<HashMap<Pubkey, u64>>>,
+    node_used: HashMap<Pubkey, /*bytes*/ u64>,
+    staked_remaining: u64,
+    unstaked_remaining: u64,
+    staked_request: u64,
+    unstaked_request: u64,
+}
+
+impl RepairBudget {
+    const INTERVAL_MS: u64 = 1_000;
+    const MAX_BYTES_PER_INTERVAL: u64 = 12_000_000;
+    const MAX_STAKED_BYTES_PER_INTERVAL: u64 = 10_000_000;
+    const MAX_NODE_BYTES_PER_INTERVAL: u64 = 1_000_000;
+
+    fn new(bank_forks: &RwLock<BankForks>) -> Self {
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        let epoch_schedule = root_bank.epoch_schedule();
+        let epoch = epoch_schedule.get_epoch(root_bank.slot());
+        let staked_nodes = root_bank.epoch_staked_nodes(epoch);
+        Self {
+            last_update: Instant::now(),
+            epoch_staked_nodes: staked_nodes,
+            node_used: HashMap::default(),
+            staked_remaining: 0,
+            unstaked_remaining: Self::MAX_BYTES_PER_INTERVAL,
+            staked_request: 0,
+            unstaked_request: 0,
+        }
+    }
+
+    fn take(&mut self, id: &Pubkey, bytes: u64) -> bool {
+        // Enforce a per-node cap on bandwith usage. Return early if the cap is
+        // hit to prevent the node from skewing the stake/unstaked capacity
+        // adjustments.
+        let node_used = self.node_used.entry(*id).or_default();
+        let requested_node_usage = node_used.saturating_add(bytes);
+        if requested_node_usage > Self::MAX_NODE_BYTES_PER_INTERVAL {
+            return false;
+        }
```

Comment:
This is tricky to enforce.
If a node is starting up and has to catch up then it will have to do a lot of repairs.
Also if there is outbound budget remaining and there is no other request, why should the request be dropped here.

I think it would be better not to add such a limit and instead respond to the requests in order of their stake.

Reply:
Hopefully a node which is catching up will still fall below this threshold given that their repair requests should be spread across multiple nodes. This would prevent a single node from consuming the entire bandwidth for the interval. This is a piece of data to be gained from metrics: the peak per-node consumption that we observe.

How do you propose ordering by stake without buffering requests over the interval?

Reply:
> Hopefully a node which is catching up will still fall below this threshold given that their repair requests should be spread across multiple nodes.

We already have seen reports that nodes not being able to catch up with the cluster if they have to do a lot of repair. This is already an outstanding issue.
Also, again, it is not clear why we should drop requests if there is outbound budget left.

> How do you propose ordering by stake without buffering requests over the interval?

at the very least we can drain the channel, then order the requests by their stake.

Reply:
On testnet I'm seeing spikes of 130000 packets/s and a load shedding drop rate > 90%. You're suggesting that we drain the channel each iteration and sort the requests by stake before load shedding?

Reply:
either that (i.e. drain then sort) or, can drain the channel and keep pushhing to a `MinMaxHeap` based on the stake:
https://docs.rs/min-max-heap/1.3.0/min_max_heap/
While keeping the heap size bounded, i.e. if it gets larger than `x`, then `pop_min`.
Once the channel is empty can respond to request by `pop_max`.

I don't know how much overhead that will add though and if it would be efficient at all.
Just suggesting some ideas here.